src/Language/DifferentialDatalog/Compile.hs

```diff
@@ -547,7 +544,51 @@ compileLib d specname rs_code = (typesLib, valueLib, mainLib)
     typedefs = vcat $ map (mkTypedef d) $ M.elems $ progTypedefs d
     -- Functions
     (fdef, fextern) = partition (isJust . funcDef) $ M.elems $ progFunctions d
-    funcs = vcat $ (map (mkFunc d) fextern ++ map (mkFunc d) fdef)
+    funcs = let ?statics = statics in vcat $ (map (mkFunc d) fextern ++ map (mkFunc d) fdef)
+
+-- This type stores the set of statically evaluated constant sub-expressions.
+-- We need to track the type of each expression, as generic functions like
+-- 'set_empty()' can have different types in different contexts.
```

Comment:
One solution would be to have an earlier pass which annotates all polymorphic functions with their type.